src/radical/pilot/utils/component.py

```diff
@@ -4,301 +4,382 @@
 #
-class Component(ru.Process):
-    """
-    This class provides the basic structure for any RP component which operates
-    on stateful things.  It provides means to:
+class ComponentManager(object):
 
-      - define input channels on which to receive new things in certain states
-      - define work methods which operate on the things to advance their state
-      - define output channels to which to send the things after working on them
-      - define notification channels over which messages with other components
-        can be exchanged (publish/subscriber channels)
+    # --------------------------------------------------------------------------
+    #
+    def __init__(self, cfg):
+
+        self._cfg  = ru.Config('radical.pilot.cmgr', cfg=cfg)
+        self._sid  = self._cfg.sid
+        self._uid  = ru.generate_id('cmgr', ns=self._sid)
+        self._uids = [self._uid]  # uids to track hartbeats for (incl. own)
+
+        self._prof = ru.Profiler(self._uid, ns='radical.pilot',
+                               path=self._cfg.path)
+        self._log  = ru.Logger(self._uid, ns='radical.pilot',
+                               path=self._cfg.path)
+
+        self._prof.prof('init2', uid=self._uid, msg=self._cfg.path)
+
+        # Every ComponentManager runs a HB pubsub bridge in a separate thread.
+        # That HB channel should be used by all components and bridges created
+        # under this CMGR.
+        bcfg = ru.Config(cfg={'channel'    : 'heartbeat',
+                              'type'       : 'pubsub',
+                              'uid'        : self._uid + '.hb',
+                              'stall_hwm'  : 1,
+                              'bulk_size'  : 0,
+                              'path'       : self._cfg.path})
+        self._hb_bridge = ru.zmq.PubSub(bcfg)
+        self._hb_bridge.start()
+
+        self._cfg.heartbeat.addr_pub = str(self._hb_bridge.addr_pub)
+        self._cfg.heartbeat.addr_sub = str(self._hb_bridge.addr_sub)
+
+        # runs a HB monitor on that channel
+        self._hb = ru.Heartbeat(uid=self.uid,
+                                timeout=self._cfg.heartbeat.timeout,
+                                interval=self._cfg.heartbeat.interval,
+                                beat_cb=self._hb_beat_cb,  # on every heartbeat
+                                term_cb=self._hb_term_cb,  # on termination
+                                log=self._log)
+
+        self._hb_pub = ru.zmq.Publisher('heartbeat',
+                                        self._cfg.heartbeat.addr_pub,
+                                        log=self._log)
+        self._hb_sub = ru.zmq.Subscriber('heartbeat',
+                                         self._cfg.heartbeat.addr_sub,
+                                         topic='heartbeat',
+                                         cb=self._hb_sub_cb, log=self._log)
+
+        # confirm the bridge being usable by listening to our own heartbeat
```

Comment:
What we do if this 'confirmation' fails?

Reply:
if heartbeats fail we die...